Bottom: f8bae1feaa568b165612f0c5073268671f3a11ba
Top:    47e54029ad559cdf565b84ba7732225a32190799
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-03-01 16:59:54 +0800

vhost: tx batching

xdp_redirect.sh shows 13.7% improvement.

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 12bcfba..90b0d5f 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -110,6 +110,7 @@ struct vhost_net_virtqueue {
 	struct vhost_net_ubuf_ref *ubufs;
 	struct ptr_ring *rx_ring;
 	struct vhost_net_buf rxq;
+	struct vring_used_elem heads[VHOST_RX_BATCH];
 };
 
 struct vhost_net {
@@ -473,6 +474,7 @@ static void handle_tx(struct vhost_net *net)
 	struct socket *sock;
 	struct vhost_net_ubuf_ref *uninitialized_var(ubufs);
 	bool zcopy, zcopy_used;
+	s16 nheads = 0;
 
 	mutex_lock(&vq->mutex);
 	sock = vq->private_data;
@@ -547,6 +549,8 @@ static void handle_tx(struct vhost_net *net)
 			atomic_inc(&ubufs->refcount);
 			nvq->upend_idx = (nvq->upend_idx + 1) % UIO_MAXIOV;
 		} else {
+			nvq->heads[nheads].id = cpu_to_vhost32(vq, head);
+			nvq->heads[nheads].len = 0;
 			msg.msg_control = NULL;
 			ubufs = NULL;
 		}
@@ -575,10 +579,13 @@ static void handle_tx(struct vhost_net *net)
 		if (err != len)
 			pr_debug("Truncated TX packet: "
 				 " len %d != %zd\n", err, len);
-		if (!zcopy_used)
-			vhost_add_used_and_signal(&net->dev, vq, head, 0);
-		else
+		if (zcopy_used)
 			vhost_zerocopy_signal_used(net, vq);
+		else if (++nheads == VHOST_RX_BATCH) {
+			vhost_add_used_and_signal_n(&net->dev, vq, nvq->heads,
+						    nheads);
+			nheads = 0;
+		}
 		vhost_net_tx_packet(net);
 		if (unlikely(total_len >= VHOST_NET_WEIGHT)) {
 			vhost_poll_queue(&vq->poll);
@@ -586,6 +593,8 @@ static void handle_tx(struct vhost_net *net)
 		}
 	}
 out:
+	if (nheads)
+		vhost_add_used_and_signal_n(&net->dev, vq, nvq->heads, nheads);
 	mutex_unlock(&vq->mutex);
 }
